blockstore/net_serve.go

```diff
@@ -0,0 +1,236 @@
+					log.Warnw("writing error response", "error", err)
+					return
+				}
+				continue
+			}
+
+			var resData [1]byte
+			if res {
+				resData[0] = cbg.CborBoolTrue[0]
+			} else {
+				resData[0] = cbg.CborBoolFalse[0]
+			}
+
+			if err := h.respond(req.ID, NRpcOK, resData[:]); err != nil {
+				log.Warnw("writing response", "error", err)
+				return
+			}
+
+		case NRpcGet:
+			if len(req.Cid) != 1 {
+				if err := h.respondError(req.ID, xerrors.New("expected request for 1 cid"), cid.Undef); err != nil {
+					log.Warnw("writing error response", "error", err)
+					return
+				}
+				continue
+			}
+
+			err := h.bs.View(ctx, req.Cid[0], func(bdata []byte) error {
+				return h.respond(req.ID, NRpcOK, bdata)
+			})
+			if err != nil {
+				if err := h.respondError(req.ID, err, req.Cid[0]); err != nil {
+					log.Warnw("writing error response", "error", err)
+					return
+				}
+				continue
+			}
+
+		case NRpcGetSize:
+			if len(req.Cid) != 1 {
+				if err := h.respondError(req.ID, xerrors.New("expected request for 1 cid"), cid.Undef); err != nil {
+					log.Warnw("writing error response", "error", err)
+					return
+				}
+				continue
+			}
+
+			sz, err := h.bs.GetSize(ctx, req.Cid[0])
+			if err != nil {
+				if err := h.respondError(req.ID, err, req.Cid[0]); err != nil {
+					log.Warnw("writing error response", "error", err)
+					return
+				}
+				continue
+			}
+
+			var resData [4]byte
+			binary.LittleEndian.PutUint32(resData[:], uint32(sz))
+
+			if err := h.respond(req.ID, NRpcOK, resData[:]); err != nil {
+				log.Warnw("writing response", "error", err)
+				return
+			}
+
+		case NRpcPut:
+			blocks := make([]block.Block, len(req.Cid))
+
+			if len(req.Cid) != len(req.Data) {
+				if err := h.respondError(req.ID, xerrors.New("cid count didn't match data count"), cid.Undef); err != nil {
+					log.Warnw("writing error response", "error", err)
+				}
+				return
+			}
+
+			for i := range req.Cid {
+				blocks[i], err = block.NewBlockWithCid(req.Data[i], req.Cid[i])
+				if err != nil {
+					log.Warnw("make block", "error", err)
+					return
+				}
+			}
+
+			err := h.bs.PutMany(ctx, blocks)
+			if err != nil {
+				if err := h.respondError(req.ID, err, cid.Undef); err != nil {
+					log.Warnw("writing error response", "error", err)
+					return
+				}
+				continue
+			}
+
+			if err := h.respond(req.ID, NRpcOK, []byte{}); err != nil {
+				log.Warnw("writing response", "error", err)
+				return
+			}
+		case NRpcDelete:
+			err := h.bs.DeleteMany(ctx, req.Cid)
+			if err != nil {
+				if err := h.respondError(req.ID, err, cid.Undef); err != nil {
+					log.Warnw("writing error response", "error", err)
+					return
+				}
+				continue
+			}
+
+			if err := h.respond(req.ID, NRpcOK, []byte{}); err != nil {
+				log.Warnw("writing response", "error", err)
+				return
+			}
+		case NRpcList:
+			if err := h.respondError(req.ID, xerrors.New("list todo"), cid.Undef); err != nil {
```

Comment:
nit: It's unexpected that one side of List/AllKeys chan is implemented completely and the other is not.

Reply:
Yeah, I wrote this mainly for retrieval, and iterating over blockstores isn't exactly needed for that. Maybe it's better to just also drop the client side of this for now instead of pretending like it's half done.